### ISSUE TITLE
[Bug] Deck Options are not saved until you go back to Deck screen

### ISSUE THREAD

**Issue by bhatushar** [ankidroid/Anki-Android#8180]
###### Reproduction Steps

1. Go to deck options
2. Modify any setting
3. Close app

###### Expected Result

The changes should persist.

###### Actual Result

The changes are not stored until you go back to the Deck screen.

**Comment by david-allison-1** [comment]
For design discussion

It's currently not possible for a user to undo changes that are made on this screen. Closing the app allows a user to back-out of changes that they've made.

It might be prudent to add in a "save" button to make this change explicit, but in this case it would still be correct to not save if the screen is closed.

**Comment by avirajrsingh** [comment]
@david-allison-1 What if we show a alert dialog to the users if they try to exit the application without saving changes? We can give options to discard or save directly in the dialog itself, if it makes sense.

**Comment by mikehardy** [comment]
There is no way to know conclusively when a user exits an app (or an app is killed etc) on Android, the shutdown / exit hooks etc are unreliable.

I'd be +1 for an explicit save button - making an action explicit is usually the right choice in my opinion

**Comment by bhatushar** [comment]
If no one else is working on it, I'd like to give it a go. I like @david-allison-1's approach of adding a "Save" button in the DeckOption activity.

**Comment by david-allison-1** [comment]
Sounds good to me, thanks!

**Comment by bhatushar** [comment]
@david-allison-1 @mikehardy I'd like some clarification.

I'm adding two buttons - Save and Cancel. Now if the user exists Deck Options by pressing the Back key, should I store the result (default behavior) or discard it?

I feel discarding it would be more intuitive but it'll also require making a lot of changes to DeckOptions class which might break something. What do you think?

**Comment by david-allison-1** [comment]
Also check with Mike. Personally, let's do this right

* Back & no changes - back
* Back & changes - `ConfirmationDialog`: Are you sure

Class: `DiscardChangesDialog` may be useful as it's a common pattern

```java
        DiscardChangesDialog
                .getDefault(this)
                .onPositive((dialog, which) -> discardChangesAndClose())
                .show();
```

**Comment by mikehardy** [comment]
I agree with @david-allison-1 there, if there is a save, there should be a cancel, and if they go back with out a cancel ask for confirm

**Comment by bhatushar** [comment]
@david-allison-1 @mikehardy I'm stuck on a problem and could use some help.

So, I found out that `DeckOptions` was saving the changes no matter what the `setResult()` parameter is. This is because `DeckOptions` was not started with `startActivityForResult()` inside `DeckPicker` class. 

I know the final modifications need to be made inside `onActivityResult()` of `DeckPicker` such that changes are saved only when the result code is ok. But I can't figure out what the code is to commit the changes.

**Comment by bhatushar** [comment]
@david-allison-1 @mikehardy Can you tell me in which code section the deck options are being committed to the database? As it is now, having `setResult(RESULT_CANCELED)` inside `DeckOptions.closeWithResult()` still saves the results.

I tried going through the parent activity `Reviewer` but couldn't find any control branch that might be responsible for storing data.

**Comment by david-allison-1** [comment]
Not too familiar with this screen

Should be:

https://github.com/ankidroid/Anki-Android/blob/7404c67eb80298b686d52de79b807c34363878df/AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java#L193

Run the code under a debugger, set a breakpoint there and see what the call stack is.

**Comment by bhatushar** [comment]
@david-allison-1 That doesn't seem to be it. 

`commit()` is called when a preference is changed but the data is saved when we exit the deck options activity. This is the tread dump when `commit()` is called:

```
"main@9108" prio=5 tid=0x2 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at com.ichi2.anki.DeckOptions$DeckPreferenceHack$Editor.commit(DeckOptions.java:219)
	  at com.ichi2.anki.DeckOptions$DeckPreferenceHack$Editor.apply(DeckOptions.java:532)
	  at android.preference.Preference.tryCommit(Preference.java:1621)
	  at android.preference.Preference.persistInt(Preference.java:1763)
	  at com.ichi2.preferences.NumberRangePreference.persistString(NumberRangePreference.java:82)
	  at android.preference.EditTextPreference.setText(EditTextPreference.java:103)
	  at com.ichi2.preferences.NumberRangePreference.setValue(NumberRangePreference.java:181)
	  at com.ichi2.preferences.NumberRangePreference.onDialogClosed(NumberRangePreference.java:62)
	  at com.ichi2.preferences.IncrementerNumberRangePreference.onDialogClosed(IncrementerNumberRangePreference.java:70)
	  at android.preference.DialogPreference.onDismiss(DialogPreference.java:456)
	  at android.app.Dialog$ListenersHandler.handleMessage(Dialog.java:1407)
	  at android.os.Handler.dispatchMessage(Handler.java:107)
	  at android.os.Looper.loop(Looper.java:214)
	  at android.app.ActivityThread.main(ActivityThread.java:7356)
	  at java.lang.reflect.Method.invoke(Method.java:-1)
	  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
	  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

Do you know anyone who might be familiar with working on `DeckOptions`?

**Comment by bhatushar** [comment]
UPDATE: I figured out how to save/discard changes, but it doesn't take effect until app restart. For example, if you make any changes and then discard them, the app will still show the changes you've made even though they're not committed to the database.

**Comment by mikehardy** [comment]
Typically that is handled as part of the return code from startActivityForResult I think - you can pass result codes that indicate to the original/calling activity "hey, you need to reload yourself in response to this return code"

The card browser does this in response to some result codes, specifically if you edit a note in the card browser when you are done editing the card browser is told to reload (to include the edits)

**Comment by bhatushar** [comment]
@mikehardy I tried playing around with `saveResult()` but it seemed the result code had no effect on the data. Changes were still persisting even after setting `RESULT_CANCELED`.

In my PR, I worked around it by undoing all the unsaved changes before closing the deck options activity.

**Comment by mikehardy** [comment]
That sounds mildly frightening. Partial work should not be committed, and then undone to prevent partial work commit. Something is wrong, and needs a real understanding and a real fix.

**Comment by bhatushar** [comment]
No, no. Partial changes **were** being committed. My PR fixes that. 

Nothing is being written to DB until the user explicitly saves.

**Comment by mikehardy** [comment]
Ah, whew! :-)

There was another place in the program that used to do the same and it was quite a mess.

I still am very suspicious of this:

>  I tried playing around with saveResult() but it seemed the result code had no effect on the data. Changes were still persisting even after setting RESULT_CANCELED.

Something is wrong here, understanding it (if you have not already) is important. If you already have that whole area handled then I misunderstand your comment and it would be good for you to post one clear comment with current status

**Comment by bhatushar** [comment]
I really tried to figure out `setResult()`. I followed the call trace using a debugger, but it didn't help either. It still remains a mystery.

But from what I could gather, the statement `mCol.getDecks().save()` is what's responsible for writing changes to DB. The save function was called in several different places inside DeckOptions.java. So, I aggregated all those calls in one place in other to avoid inconsistencies.
